### DISCUSSION TITLE
Per branch environment variables

### DISCUSSION THREAD

**Discussion by matteobad** [coollabsio/coolify#4237] [Feature Requests]
Coolify now offers preview environments with custom environment variables, while this is great I think is just the first step to actually enable a complete workflow. Let's say that I deploy a Next.js + Postgres application as of right now I have two options:

1. Docker compose deploy connected via docker internal network
2. Next.js application + Postgres service-connected vai ENV

In the first option, I could have preview deployments using a DATABASE_URL on the internal docker network. On the second option, I would have to spin up another Postgres and the appropriate DATABASE_URL in the preview envs, which means no more than 1 {PREVIEW}_DATABASE_URL at a time.

On Vercel on the other hand we can set preview ENVs per branch like so:
![image](https://github.com/user-attachments/assets/1fcb2752-fa9e-4ab6-b8c0-2e16bda4608c)

With this approach, we could also integrate into the workflow DB branching tools like Neon or Planetscale and let each preview env connect to its services.

I'm I missing something? I'll be happy to further explore this :)

**Comment by janwiebe-jump** [comment]
I tried to use templates like {{pr_id}} in the preview section of the environment variables of my project based on docker compose.
Unfortunately they are not substituted during the build into the container's environment.

**Comment by Arnaud-IFEA** [comment]
I don't understand either how to make a docker compose work with Preview Deployment...

if i have 2 services , one front and one db :

```yaml
db:
  image: postgres_whatever
.../...
myapp:
 .../...
```
if i use the internal docker hostname `db` to connect from `app`, it works for the main deployment. But if i use preview deployment, the service names are changed ( `db-pr-50` for example) and there is no `SERVICE_*` magic variable to get the new name... and so it fails...

I don't understand why the service names are changed, it's a new network so there is no problem to keep the same service name for every pr...
Any idea or workaround ?

**Reply by djsisson** [reply]
i've no idea why this is the case - i can only assume it might be purely to do with how its saved in the coolify db and each service needs its own row, and servicename has to be unique (no idea just guessing)

https://github.com/coollabsio/coolify/blob/f59a68993a2ef8b848fc55b2cbc639944e098d27/bootstrap/helpers/shared.php#L2898-L2902

since as you say it breaks your networking using the old service name

as a work around you can add a network alias, 

```yaml
services:
  app:
    networks:
      - customnetwork
  db:
    networks:
      customnetwork:
        aliases:
          - db
networks:
  customnetwork: {}
```

coolify would still add another network and add traefik to it, but this would be your internal compose network, and you can access the db using the alias you gave and so the change of service name wouldn't matter
